### ISSUE TITLE
validate-superclass methods need to be avl-when 9compile load eval) to allow compilation without loading first

### ISSUE THREAD

**Issue by Yehouda** [Shirakumo/maiden#13]
https://github.com/Shirakumo/maiden/blob/ba92d948b8c5548dabbe38b6bf52f90a2d0a100d/consumer.lisp#L11

These **validate-cuperclass** methods are require down in the file, so need to be inside **(eval-when (compile load eval)..**. Without it compilation (without loading first) errors when trying to imherit **named-entity** from **consumer**

Even better have a spearate file for the metaclass **consumer-class** and all its methods.

**Comment by Shinmera** [comment]
I don't think that's true? Neither SBCL, CCL, nor ECL have an issue with it either.

**Comment by Yehouda** [comment]
If that is true, that is a bug in SBCL, CCL and ECL. 

https://www.lispworks.com/documentation/HyperSpec/Body/m_defmet.htm

Penultimate paragraph in the description:

"In particular, the method are not installed for invocation during compilation."

**Comment by Shinmera** [comment]
I don't think so. The method is not necessary to be installed at that point in the file.

**Comment by Yehouda** [comment]
So how it is supposed to work when it is called later in the file?
It needs to be used when it defines the class **named-entity** because **defclass** does need to happen at compile time. 
https://www.lispworks.com/documentation/HyperSpec/Body/m_defcla.htm  Last paragraph of the description.

**Comment by Shinmera** [comment]
No, defclass does not in fact need to happen at compile time. Only the name needs to be recognised at compile time. The class does not.

**Comment by Shinmera** [comment]
Anyway are you like, just, looking at the source code and thinking of problems? Or do you have an actual bug to report that you can see in real life, because if not, you're wasting my time.

**Comment by Yehouda** [comment]
It says:
"The compiler must make the class definition available to be returned by find-class when its environment argument is a value received as the environment parameter of a macro. "
So the compiler has to make a class definition, not only a name. 

I download all the quicklisp systems, compile them all and if can identify tests run them, and check for any problems.

**Comment by Shinmera** [comment]
What is the actual bug you have. Post a fucking backtrace. An error description. Your implementation. Something. Not just your guesses as to what's to be done about something that may or may not be an issue.

**Comment by Yehouda** [comment]
It gave an error like that:
```

**++++ Error in (DEFCLASS MAIDEN:CONSUMER): 
  #<STANDARD-CLASS NAMED-ENTITY 82E054C0BB> is an invalid superclass of #<CONSUMER-CLASS CONSUMER 801005E47B>.

Error: #<STANDARD-CLASS NAMED-ENTITY 82E054C0BB> is an invalid superclass of #<CONSUMER-CLASS CONSUMER 801005E47B>.
   1 Skip compiling form.
```

Couldn't bother about a backtrace. I just went and fixed the my copy, so it now does run.

**Comment by Shinmera** [comment]
I still don't know which implementation this is supposed to be

**Comment by Yehouda** [comment]
LispWorks

**Comment by Harleqin** [comment]
I am not completely convinced that the assumption in the initial posting here about the CL spec is correct.  Even if it were correct, I think that implementations are at least _allowed_ to handle this situation as intended (as SBCL etc. seem to do), see the `defclass` spec page (look for the context of this sentence):

> The object system can be extended to cover situations where these rules are not obeyed.

The mentioned rules are for `standard-class`es, but it is not specified that this they _must not_ hold for other kinds of classes.

In any case, I think it would have been better for smooth communication to start this issue with a bug report, or to at least give one when asked.
